velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -0,0 +1,241 @@
+              << iteration << " (seed: " << currentSeed_ << ")";
+
+    auto signatureWithStats = pickSignature();
+    signatureWithStats.second.numRuns++;
+
+    auto signature = signatureWithStats.first;
+    stats_.functionNames.insert(signature.name);
+
+    const bool customVerification =
+        customVerificationFunctions_.count(signature.name) != 0;
+    const bool requireSortedInput =
+        orderDependentFunctions_.count(signature.name) != 0;
+
+    std::vector<TypePtr> argTypes = signature.args;
+    std::vector<std::string> argNames = makeNames(argTypes.size());
+
+    auto call = makeFunctionCall(signature.name, argNames, false);
+
+    std::vector<std::string> sortingKeys;
+    // 50% chance without order-by clause.
+    if (vectorFuzzer_.coinToss(0.5)) {
+      sortingKeys = generateSortingKeys("s", argNames, argTypes);
+    }
+    auto partitionKeys = generateKeys("p", argNames, argTypes);
+    auto input = generateInputDataWithRowNumber(argNames, argTypes, signature);
+    // If the function is order-dependent, sort all input rows by row_number
+    // additionally.
+    if (requireSortedInput) {
+      sortingKeys.push_back("row_number");
+      ++stats_.numSortedInputs;
+    }
+
+    logVectors(input);
+
+    bool failed = verifyWindow(
+        partitionKeys,
+        sortingKeys,
+        {call},
+        input,
+        customVerification,
+        FLAGS_enable_window_reference_verification);
+    if (failed) {
+      signatureWithStats.second.numFailed++;
+    }
+
+    LOG(INFO) << "==============================> Done with iteration "
+              << iteration;
+
+    if (persistAndRunOnce_) {
+      LOG(WARNING)
+          << "Iteration succeeded with --persist_and_run_once flag enabled "
+             "(expecting crash failure)";
+      exit(0);
+    }
+
+    reSeed();
+    ++iteration;
+  }
+
+  stats_.print(iteration);
+  printSignatureStats();
+}
+
+void WindowFuzzer::go(const std::string& /*planPath*/) {
+  // TODO: allow running window fuzzer with saved plans and splits.
+  VELOX_NYI();
+}
+
+void WindowFuzzer::updateReferenceQueryStats(
+    AggregationFuzzerBase::ReferenceQueryErrorCode ec) {
+  if (ec == ReferenceQueryErrorCode::kReferenceQueryFail) {
+    ++stats_.numReferenceQueryFailed;
+  } else if (ec == ReferenceQueryErrorCode::kReferenceQueryUnsupported) {
+    ++stats_.numVerificationNotSupported;
+  }
+}
+
+bool WindowFuzzer::verifyWindow(
+    const std::vector<std::string>& partitionKeys,
+    const std::vector<std::string>& sortingKeys,
+    const std::vector<std::string>& functionCalls,
+    const std::vector<RowVectorPtr>& input,
+    bool customVerification,
+    bool enableWindowVerification) {
+  std::stringstream frame;
+  VELOX_CHECK(!partitionKeys.empty());
+  frame << "partition by " << folly::join(", ", partitionKeys);
+  if (!sortingKeys.empty()) {
+    frame << " order by " << folly::join(", ", sortingKeys);
+  }
+  // TODO: allow randomly generated frames.
+  auto plan =
```

Comment:
Would you consider adding a query plan with TableScan instead of Values node and verify that both plans return same results. Maybe also, add a plan with Values node that uses flat inputs.